nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,372 @@
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!filter_mask_) {
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Received filter mask from %s topic.", mask_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "New filter mask arrived from %s topic. Updating old filter mask.",
+      mask_topic_.c_str());
+    filter_mask_.reset();
+  }
+
+  filter_mask_ = msg;
+  mask_frame_ = msg->header.frame_id;
+}
+
+bool SpeedFilter::transformPose(
+  const geometry_msgs::msg::Pose2D & pose,
+  geometry_msgs::msg::Pose2D & mask_pose) const
+{
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (mask_frame_ != global_frame_) {
+    // Filter mask and current layer are in different frames:
+    // Prepare transform from current layer frame (global_frame_) to mask frame
+    geometry_msgs::msg::TransformStamped transform;
+    tf2::Transform tf2_transform;
+    tf2_transform.setIdentity();  // initialize by identical transform
+    try {
+      transform = tf_->lookupTransform(
+        mask_frame_, global_frame_, tf2::TimePointZero,
+        transform_tolerance_);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR(
+        node->get_logger(),
+        "SpeedFilter: failed to get costmap frame (%s) "
+        "transformation to mask frame (%s) with error: %s",
+        global_frame_.c_str(), mask_frame_.c_str(), ex.what());
+      return false;
+    }
+    tf2::fromMsg(transform.transform, tf2_transform);
+
+    // Transform (pose.x, pose.y) point from global_frame_
+    // to mask_pose point in mask_frame_
+    tf2::Vector3 point(pose.x, pose.y, 0);
+    point = tf2_transform * point;
+    mask_pose.x = point.x();
+    mask_pose.y = point.y();
+  } else {
+    // Filter mask and current layer are in the same frame:
+    // Just use pose coordinates
+    mask_pose = pose;
+  }
+
+  return true;
+}
+
+bool SpeedFilter::worldToMask(double wx, double wy, unsigned int & mx, unsigned int & my) const
+{
+  double origin_x = filter_mask_->info.origin.position.x;
+  double origin_y = filter_mask_->info.origin.position.y;
+  double resolution = filter_mask_->info.resolution;
+  unsigned int size_x = filter_mask_->info.width;
+  unsigned int size_y = filter_mask_->info.height;
+
+  if (wx < origin_x || wy < origin_y) {
+    return false;
+  }
+
+  mx = std::round((wx - origin_x) / resolution);
+  my = std::round((wy - origin_y) / resolution);
+  if (mx >= size_x || my >= size_y) {
+    return false;
+  }
+
+  return true;
+}
+
+void SpeedFilter::process(
+  nav2_costmap_2d::Costmap2D & /*master_grid*/,
+  int /*min_i*/, int /*min_j*/, int /*max_i*/, int /*max_j*/,
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
```

Comment:
Please just store logger in logger_ and clock in `clock_` and use that so you don't need to get the node and lock it every time just to get a logger/clock. That's the new design paradigm we have for all the plugins after the last set of PRs to fix shutdown.

Reply:
Regarding new paradigm: how about cases where we need to call `node`'s basic functions: e.g. `node->create_subscription()`.
Does the paradigm implies to still use 
```
  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
  if (!node) {
    throw std::runtime_error{"Failed to lock node"};
  }

```
local pointer each time when we need to make a subscription, publisher, or process any other action with a `node` ?
Or maybe is it better to add `shared_node_` as a plugin member in the paradigm (not sure if this will not lead to memory leaks again, what we initially wanted to fix away from)?

Reply:
Done in https://github.com/ros-planning/navigation2/pull/2074/commits/3465c90415afd11418f98fc01bbc9386afa853ea